feat: Editorial review: Attribution reporting API (#33172)
mdn/content: files/en-us/web/api/attribution_reporting_api/generating_reports/index.md

---
title: Generating attribution reports
slug: Web/API/Attribution_Reporting_API/Generating_reports
page-type: guide
status:
  - experimental
---

{{SeeCompatTable}}{{DefaultAPISidebar("Attribution Reporting API")}}

> This article explains how reports are generated — both attribution reports and debug reports — and how you can control the generated reports. This includes handling noise, prioritizing reports, filtering reports, and generating debug reports.

Reviewer: Link to API top page IMO

Author: added

## Basic process

> When a match occurs between a trigger and a source, the browser generates a report and sends it via an uncredentialed [`POST`](/en-US/docs/Web/HTTP/Methods/POST) request to a specific endpoint on the reporting origin:

Reviewer: We should say here at what point/how the browser determined the reporting origin.

Author: Just below this, it says "The `<reporting-origin>` will be same-origin with the one that registered the source and trigger." Does this not answer your question?

Reviewer: Doh

```json
{
  "attribution_destination": "https://advertiser.example",
  "source_event_id": "412444888111012",
  "trigger_data": "412444888111012",
  "report_id": "123e4567-e89b-12d3-a456-426614174000",
  "source_type": "navigation",
  "randomized_trigger_rate": 0.34,
  "scheduled_report_time": "1692255696",
  "source_debug_key": 647775351539539,
  "trigger_debug_key": 647776891539539
}
```

The properties are as follows:

- `"attribution_destination"`
  - : A string, or an array of 2–3 strings, depending on whether the source was registered with multiple destinations or not. These strings represent the attribution [`"destination"`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#destination) site(s) set in the source registration via the associated {{httpheader("Attribution-Reporting-Register-Source")}} response header.
- `"source_event_id"`
  - : A string representing the attribution source ID. This is equal to the `source_event_id` set in the source registration (via the associated {{httpheader("Attribution-Reporting-Register-Source")}} response header).
- `"trigger_data"`

>   - : A string representing coarse data originating from the attribution trigger, set in the trigger registration (via the associated {{httpheader("Attribution-Reporting-Register-Trigger")}} response header).

Reviewer: Good to link to the specific data in the json.

Author: Added specific links, in all cases where there were not specific links.

- `"report_id"`
  - : A string representing a [Universally Unique Identifier (UUID)](/en-US/docs/Glossary/UUID) for this report, which can be used to prevent duplicate counting.
- `"source_type"`
  - : A string equal to either `"navigation"` or `"event"`, which respectively indicate whether the associated attribution source is [navigation-based](/en-US/docs/Web/API/Attribution_Reporting_API/Registering_sources#navigation-based_attribution_sources), or [event-based](/en-US/docs/Web/API/Attribution_Reporting_API/Registering_sources#event-based_attribution_sources).
- `"randomized_trigger_rate"`

>   - : A random number between 0 and 1 indicating how often noise is applied for this particular source configuration.

Reviewer: Noise is mentioned here and in other places. No explanation of why/when/how. Link to explanation please.

Reviewer: I see https://pr33172.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/Attribution_Reporting_API/Generating_reports#how_noise_is_added_to_reports - link would be noice :-)

Author: links added

- `"scheduled_report_time"`

>   - : A string representing the number of seconds from the Unix Epoch until the browser initially scheduled the report to be sent (to avoid noise around offline devices reporting late).

Reviewer: Is this the same noise as above?

Author: No, this is noise that you can't control, as opposed to controlled noise added to improve privacy and mitigate fingerprinting. To avoid confusion, I've changed the wording here to "to avoid inaccuracies as a result of offline devices reporting late"

A typical aggregatable report might look like this:

```json
{
  "shared_info": "{\"api\":\"attribution-reporting\",\"attribution_destination\":\"https://advertiser.example\",\"report_id\":\"123e4567-e89b-12d3-a456-426614174000\",\"reporting_origin\":\"https://reporter.example\",\"scheduled_report_time\":\"1692255696\",\"source_registration_time\":\"1692230400\",\"version\":\"3\"}",
  "aggregation_service_payloads": [
    {
      "payload": "[base64-encoded HPKE encrypted data readable only by the aggregation service]",
      "key_id": "[string identifying public key used to encrypt payload]",
      "debug_cleartext_payload": "[base64-encoded unencrypted payload]"
    }
  ],
  "aggregation_coordinator_origin": "https://publickeyservice.aws.privacysandboxservices.com",
  "source_debug_key": 647775351539539,
  "trigger_debug_key": 647776891539539
}
```

The properties are as follows:

- `"shared_info"`

>   - : This is a serialized JSON object providing information that an aggregation service will use to put together a summary report. This data is [encrypted](/en-US/docs/Glossary/Encryption) using [AEAD](https://en.wikipedia.org/wiki/Authenticated_encryption) to prevent tampering. The following properties are represented in the serialized string:

Reviewer: Just confirming, to present tampering or for privacy? Usually digitally sign for tamper protection.

Author: tampering. The whole API is set up for better privacy while implementing attribution reporting use cases.

    - `"attribution_destination"`
      - : A string representing the attribution `destination` URL set in the source registration (via the associated {{httpheader("Attribution-Reporting-Register-Source")}} response header).
    - `"report_id"`
      - : A string representing a [Universally Unique Identifier (UUID)](/en-US/docs/Glossary/UUID) for this report, which can be used to prevent duplicate counting.
    - `"reporting_origin"`
      - : The origin that triggered the report generation.
    - `"scheduled_report_time"`
      - : A string representing the number of seconds from the Unix Epoch until the browser initially scheduled the report to be sent (to avoid noise around offline devices reporting late).
    - `"source_registration_time"`
      - : A string representing the number of seconds from the Unix Epoch until the attribution source was registered, rounded down to a whole day.
    - `"version"`
      - : A string representing the version of the API used to generate the report.
- `"aggregation_service_payloads"`

>   - : An array of objects representing payload objects containing the histogram contributions used by the aggregation service to assemble the data contained in the report (EDITORIAL: I'M GUESSING HERE; I DON'T UNDERSTAND WHAT THIS DOES). Currently, only a single payload is supported per report, configured by the browser. In the future multiple, customizable payloads may be supported (EDITORIAL: AGAIN, GUESSING, BUT THIS IS WHAT THE EXPLAINER IMPLIES). Each payload object can contain the following properties:

Reviewer: EDITORIAL MARKUP HERE!

Author: oops, removed!

            "bucket": <bucket, encoded as a 16-byte (i.e. 128-bit) big-endian bytestring>,
            "value": <value, encoded as a 4-byte (i.e. 32-bit) big-endian bytestring>
          }, ...]
        }
        ```

    - `"key_id"`
      - : A string identifying the public key used to encrypt the payload.
    - `"debug_cleartext_payload"` {{optional_inline}}
      - : Optional debugging information.

- `"aggregation_coordinator_origin"`
  - : The deployment option for the aggregation service.
- `"source_debug_key"` {{optional_inline}}
  - : A 64 bit unsigned integer representing the debugging key for the attribution source. This mirrors the value set in the associated {{httpheader("Attribution-Reporting-Register-Source")}} header's `"debug_key"` field.
- `"trigger_debug_key"` {{optional_inline}}
  - : A 64 bit unsigned integer representing the debugging key for the attribution trigger. This mirrors the value set in the associated {{httpheader("Attribution-Reporting-Register-Trigger")}} header's `"debug_key"` field.

## How noise is added to reports

Noise plays a significant role in both event-level and summary reports as a method of protecting user privacy.

> The [`source_event_id`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#source_event_id) contained in the `Attribution-Reporting-Register-Source` header is limited to 64 bits of information to enable uniquely identifying an interaction (such as an ad click).

Reviewer: "is limited to 64 bits of information to enable uniquely identifying an" - _limited_ and _enable_ do not seem to go together here.

Author: Yeah, this phrasing is a bit weird. Changed to:

> The [`source_event_id`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#source_event_id) contained in the `Attribution-Reporting-Register-Source` header is a base-10-formatted 64-bit unsigned integer that can uniquely identify an interaction (such as an ad click).

> The trigger-side data is therefore limited quite strictly, by limiting the amount of data and by applying noise to the data. By default, [navigation sources](/en-US/docs/Web/API/Attribution_Reporting_API/Registering_sources#navigation-based_attribution_sources) will be limited to only 3 bits of [`trigger_data`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#trigger_data) (the values 0 through 7), while [event sources](/en-US/docs/Web/API/Attribution_Reporting_API/Registering_sources#event-based_attribution_sources) will be limited to only 1 bit (the values 0 through 1). The `trigger_data` and [`trigger_data_matching`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#trigger_data_matching) fields can be customized to vary the amount of noise applied. You could set a smaller range of trigger data to reduce the amount of noise applied, or a larger range to increase the noise:

Reviewer: "data is _therefore_ limited quite strictly," - there is no relationship between the points, so there is no "therefore"

Reviewer: This confuses me utterly. 

Firstly, I can see that `"trigger_data": "412444888111012",` in the for source, which is way more than 0, 1, etc.. [(https://pr33172.content.dev.mdn.mozit.cloud/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Trigger#registering_a_trigger_for_an_event-level_report)](https://pr33172.content.dev.mdn.mozit.cloud/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#registering_a_source_for_an_event-level_report).
So the statements above seem nonseniscal.

I think we're saying that a source is registered with some `trigger_data`, which is a small array of integers. A trigger will also be registered with some `trigger_data`. 
The docs say that by default this data is `[0, 1, 2, 3, 4, 5, 6, 7]` for navigation bases sources: https://pr33172.content.dev.mdn.mozit.cloud/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#trigger_data and [0, 1] for event triggers.

I "guess" that the source and trigger trigger_data are compared on a byte by byte basis. They need to match either exactly or on some other basis. The other basis is required because noise might be added to the trigger_data sent with the trigger. So then you'd compare differently to decide if there is an actual match.

I don't think I can review this as it is.

Author: I don't understand it either. @apasel422, can you help us clear up this confusion?

Reviewer 2: Does my comment in https://github.com/mdn/content/pull/33172#discussion_r1589348378 clarify this?

Author: Apologies for the delay on this one, @apasel422. I got sucked into other things. I can return to this tomorrow. I'll let you know if anything needs further clarification.

Author: OK, I've fixed the `trigger_data` value on the `Attribution-Reporting-Register-Trigger` page — it is now set to 4, like it is on the "Registering triggers" page.

I've also had a go at improving the description of the `modulus` setting of `trigger_data_matching`. Does this make more sense?

> - `"modulus"`: In this case, the following calculation is performed — `d % allowedValues.size` — where `d` is the `"trigger_data"` from the trigger, and `allowedValues` is the sequence of values in the source's `"trigger_data"` array. If the result of this calculation matches a value in the source's `"trigger_data"` array, the match is a success. In such a case, the value will always match, unless `allowedValues` is empty. This mode exists primarily for backwards compatibility with the API's behavior before exact was introduced, and as such, you'd be unlikely to use it. 
>     
>  **Note:** If `"modulus"` is used, the source's `"trigger_data"` must form a contiguous sequence of integers starting at 0. If the trigger data does not form such a sequence, an error occurs.

It still sounds bizarre to me, and I don't understand why you'd want to use this approach, but I can understand **what** it is doing now.

However, one point of confusion to clear up — there's a little note at the bottom of this description that says:

> If not specified, `"trigger_data_matching"` defaults to `"modulus"`.

Is this true, @apasel422? I must have read this somewhere, but it seems very strange to me that the default setting would be the weird, confusing legacy value is the default. This also contradicts my claim in the text above that you'd be unlikely to use it.

Reviewer: I agree. This is really helpful thanks.

```json
{
  ...,
  // The following setting would reduce noise for a navigation
  // source (default 0–7), but increase noise for an event
  // source (default 0–1)
  "trigger_data": [0, 1, 2, 3, 4]
}
```

> Noise will be applied to whether a source will be reported truthfully. When an attribution source is registered, the browser will perform one of the following steps given a probability `p` (`p` is a small percentage, which can vary based on implementation, source type, and configuration):

Reviewer: "Noise will be applied to whether a source will be reported truthfully" - Is that correct English? I can't parse it.

Author: I've just deleted that sentence; it makes more sense without it.

## Debug reports

You can enable debug reports to return troubleshooting information about your attribution reports. These can, for example, be used to check that your setup is working properly and understand gaps in measurement results between your old cookie-based implementation and your new Attribution Reporting implementation. Debug reports are sent immediately; they are not subject to the same scheduling as event-level and summary reports.

There are two different types of debug report:

- **Success debug reports** track successful generation of a specific attribution report. Success debug reports are generated and sent as soon as the corresponding trigger is registered.
- **Verbose debug reports** give you more visibility into the attribution source and attribution trigger events associated with an attribution report. They enable you to ensure that sources were registered successfully, or track missing reports and determine why they're missing (for example due to failure in source or trigger event registration or failure when sending or generating the report). Verbose debug reports are sent immediately upon source or trigger registration.

> > **Note:** To use debug reports, the reporting origin needs to set a cookie. If the origin configured to receive reports is a third party, this cookie will be a [third-party cookie](/en-US/docs/Web/Privacy/Third-party_cookies), which means that debug reports will not be available in browsers where third-party cookies are disabled/not available.

Reviewer: Flaw is reporting a broken link

Author: Yup, this is currently broken, but it won't be, as soon as https://github.com/mdn/content/pull/33138 is merged.

If neither of these fields are specified, the report window falls back to the following defaults:

- For [event-based sources](/en-US/docs/Web/API/Attribution_Reporting_API/Registering_sources#event-based_attribution_sources), the default report window ends at the source's expiry, which is set in the `Attribution-Reporting-Register-Source` [`"expiry"`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#expiry) field. This defaults to 30 days after registration if not explicitly set.
- For [navigation-based sources](/en-US/docs/Web/API/Attribution_Reporting_API/Registering_sources#navigation-based_attribution_sources), the default report windows are 2 days, 7 days, and the source's `"expiry"`.

See [Custom report windows](https://developer.chrome.com/docs/privacy-sandbox/attribution-reporting/custom-report-windows) for more details.

Once an event-level report is received at the appropriate endpoint, how the data is processed, stored, and displayed is completely up to the developer. A typical event-level report might look like this:

```json
{
  "attribution_destination": "https://advertiser.example",
  "source_event_id": "412444888111012",

>   "trigger_data": "412444888111012",

Reviewer: Really - shouldn't this be the same array as trigger's trigger_data?

Author: Crap, yeah, I fixed that.

  "report_id": "123e4567-e89b-12d3-a456-426614174000",
  "source_type": "navigation",
  "randomized_trigger_rate": 0.34,
  "scheduled_report_time": "1692255696",
  "source_debug_key": 647775351539539,
  "trigger_debug_key": 647776891539539
}
```

The properties are as follows:

- `"attribution_destination"`
  - : A string, or an array of 2–3 strings, depending on whether the source was registered with multiple destinations or not. These strings represent the attribution [`"destination"`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#destination) site(s) set in the source registration via the associated {{httpheader("Attribution-Reporting-Register-Source")}} response header.
- `"source_event_id"`
  - : A string representing the attribution source ID. This is equal to the [`"source_event_id"`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#source_event_id) set in the source registration (via the associated {{httpheader("Attribution-Reporting-Register-Source")}} response header).
- `"trigger_data"`

>   - : A string representing coarse data originating from the attribution trigger, set in the trigger registration (the [`"trigger_data"`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Trigger#trigger_data) set via the associated {{httpheader("Attribution-Reporting-Register-Trigger")}} response header).

Reviewer: So this is the string that was in the trigger, but we (perhaps) know that only one of the bits of data was actually matched. Is this "course data" because we don't want to say what we matched? 
I guess I'm saying "what is useful about this data". Why is it "coarse data"?

Author: I've just got rid of "coarse". I think it is "coarse" because it measures a general action such as "purchase", rather than something specific such as "100 toothbrushes were purchased by Mr Jones of Conneticut". But I think it confuses more than it helps here.

- `"source_debug_key"` {{optional_inline}}
  - : A 64 bit unsigned integer representing the debugging key for the attribution source. This mirrors the value set in the associated {{httpheader("Attribution-Reporting-Register-Source")}} header's [`"debug_key"`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#debug_key) field. See [Debug reports](#debug_reports) for more information.
- `"trigger_debug_key"` {{optional_inline}}
  - : A 64 bit unsigned integer representing the debugging key for the attribution trigger. This mirrors the value set in the associated {{httpheader("Attribution-Reporting-Register-Trigger")}} header's `"debug_key"` field. See [Debug reports](#debug_reports) for more information.

## How noise is added to reports

Noise is added to reports in order to obscure the output associated with a source, in order to protect user privacy. The exact source data cannot be identified and attributed back to individual users, but the overall patterns taken from the data will still provide the same meaning.

In the case of event-level reports, this is done using a randomized response algorithm, which works like so:

> 1. When an attribution source is stored, the browser generates a list of all possible sets of reports that could originate from the source's configuration (including the set consisting of no reports).

Reviewer: What's the list of reports? I mean is this "an event report", "no report", "summary report". What reports would I get out of this definition?

```json
{
    source_event_id: "412444888111012",
    destination: "https://advertiser.example",
    expiry: "604800",
    priority: "100",
    debug_key: "122939999",
    event_report_window: "86400",
   "trigger_data": [0, 1, 2, 3, 4],
   // Could also be "modulus"
   "trigger_data_matching": "exact",
  }
```

Author: I am not sure. I think it is "0 matched with trigger", "1 matched with trigger" ... up to "no matches with trigger", but let's ask @apasel422 to confirm this.

Reviewer: My intuition is that you are right but we need to confirm that nothing else affects the report space, and also 
- whether we **want** to modify this as a developer in order to change the noise
- Why the navigation and event sources have different defaults for this array.

See my note: https://github.com/mdn/content/pull/33172

> 2. In a small percentage of cases, the browser prevents the source from being attributed and instead picks a random member of that list to use as the source's reports. The probability of this happening is based on the size of that list, the browser's implementation-specific privacy parameters, and the source's chosen [`"event_level_epsilon"`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#event_level_epsilon).

Reviewer: So this tells me my assumption above is wrong, but not "what might differ between reports". I'm just trying to understand how the report might differ from what the actual attribution was, as that will help me understand the value of this.

Author: I'm not sure what your above assumption is, or what you think is needed here.

Reviewer: @chrisdavidmills Ignore this, if the point above gets answered this will be answered.

> Trigger data is limited quite strictly. By default, [navigation sources](/en-US/docs/Web/API/Attribution_Reporting_API/Registering_sources#navigation-based_attribution_sources) will be limited to only 3 bits of [`"trigger_data"`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#trigger_data) (the values 0 through 7), while [event sources](/en-US/docs/Web/API/Attribution_Reporting_API/Registering_sources#event-based_attribution_sources) will be limited to only 1 bit (the values 0 through 1). The `"trigger_data"` and [`"trigger_data_matching"`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#trigger_data_matching) fields can be customized to vary the amount of noise applied. You could set a smaller range of trigger data to reduce the amount of noise applied, or a larger range to increase the noise:

Reviewer: ```suggestion
Trigger data is limited quite strictly. By default, [navigation sources](/en-US/docs/Web/API/Attribution_Reporting_API/Registering_sources#navigation-based_attribution_sources) will be limited to only 3 bits of [`"trigger_data"`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#trigger_data) (the values 0 through 7), while [event sources](/en-US/docs/Web/API/Attribution_Reporting_API/Registering_sources#event-based_attribution_sources) will be limited to only 1 bit (the values 0 and 1). The `"trigger_data"` and [`"trigger_data_matching"`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#trigger_data_matching) fields can be customized to vary the amount of noise applied. You could set a smaller range of trigger data to reduce the amount of noise applied, or a larger range to increase the noise:
```

> Trigger data is limited quite strictly. By default, [navigation sources](/en-US/docs/Web/API/Attribution_Reporting_API/Registering_sources#navigation-based_attribution_sources) will be limited to only 3 bits of [`"trigger_data"`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#trigger_data) (the values 0 through 7), while [event sources](/en-US/docs/Web/API/Attribution_Reporting_API/Registering_sources#event-based_attribution_sources) will be limited to only 1 bit (the values 0 through 1). The `"trigger_data"` and [`"trigger_data_matching"`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#trigger_data_matching) fields can be customized to vary the amount of noise applied. You could set a smaller range of trigger data to reduce the amount of noise applied, or a larger range to increase the noise:

Reviewer: > The `"trigger_data"` and [`"trigger_data_matching"`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#trigger_data_matching) fields can be customized to vary the amount of noise applied. You could set a smaller range of trigger data to reduce the amount of noise applied, or a larger range to increase the noise:

How does that match up with the idea that a trigger data represents a particular type of trigger "user presses buy button" etc?

Author: As per previous comment, I've deleted this bit (in a slightly rage-quit kind of way ;-) )

      - : Optional debugging information.

- `"aggregation_coordinator_origin"`
  - : The deployment option for the aggregation service.
- `"source_debug_key"` {{optional_inline}}
  - : A 64 bit unsigned integer representing the debugging key for the attribution source. This mirrors the value set in the associated {{httpheader("Attribution-Reporting-Register-Source")}} header's [`"debug_key"`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#debug_key) field. See [Debug reports](#debug_reports) for more information.
- `"trigger_debug_key"` {{optional_inline}}
  - : A 64 bit unsigned integer representing the debugging key for the attribution trigger. This mirrors the value set in the associated {{httpheader("Attribution-Reporting-Register-Trigger")}} header's `"debug_key"` field. See [Debug reports](#debug_reports) for more information.

## How noise is added to reports

Noise is added to reports in order to obscure the output associated with a source, in order to protect user privacy. The exact source data cannot be identified and attributed back to individual users, but the overall patterns taken from the data will still provide the same meaning.

> This is done using a randomized response algorithm, which works like so:

Reviewer 2: We may want to clarify that this is only for event-level. Aggregatable does not use this mechanism.

Author: OK, updated to "In the case of event-level reports, this is done..." in next commit.

1. When an attribution source is stored, the browser generates a list of all possible sets of reports that could originate from the source's configuration (including the set consisting of no reports).

> 2. In a small percentage of cases, the browser prevents the source from being attributed and instead picks a random member of that report set to use as the source's reports. The probably of this happening is based on the number of reports in that set, the browser's implementation-specific privacy parameters, and the source's chosen [`"event_level_epsilon"`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#event_level_epsilon).

Reviewer 2: ```suggestion
2. In a small percentage of cases, the browser prevents the source from being attributed and instead picks a random member of that list to use as the source's reports. The probability of this happening is based on the size of that list, the browser's implementation-specific privacy parameters, and the source's chosen [`"event_level_epsilon"`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#event_level_epsilon).
```

Typical settings in the {{httpheader("Attribution-Reporting-Register-Source")}} header might look like so:

```json
{
  ...,
  // The following setting would reduce noise for a navigation
  // source (default 0–7), but increase noise for an event
  // source (default 0–1)
  "trigger_data": [0, 1, 2, 3, 4],
  // Could also be "modulus"
  "trigger_data_matching": "exact",
  ...,
}
```

> A matching {{httpheader("Attribution-Reporting-Register-Trigger")}} could containing the following:

Reviewer 2: ```suggestion
A matching {{httpheader("Attribution-Reporting-Register-Trigger")}} could contain the following:
```

```json
{
  ...,
  "event_trigger_data": [
    {
      // The value 4 is contained in the source data, therefore a match is possible
      "trigger_data": "4",
      "priority": "1000000000000",

>       "deduplication_key": "2345698765",

Reviewer 2: We probably don't need to show the `priority` and `deduplication_key` values in this example, as they are irrelevant to matching/noise.

Author: OK, removed in next commit

    },
  ],
  ...,
}
```

It is still possible that a match may not occur, based on the randomized response algorithm described above.

> > **Note:** Noise values are randomly drawn from the same specific probability distribution, regardless of whether the number of aggregated values is low or high. As a result, the noise will have a lower impact on larger values, and you are advised to scale up smaller values before sending the data through the aggregation service, and scale them back down when analysing the aggregated report data. You need to make sure that your scaling doesn't cause your data to exceed the contribution budget.

Reviewer 2: This seems to be mixing discussion of aggregatable with the preceding text, which is eent-level-only.

Author: OK, I've removed it.

I did wonder whether we should add some information on how noise works in summary reports, but instead I've just moved the link to the noise in summary reports article up a bit:

> For information on how noise works in summary reports, see [Understanding noise in summary reports](https://developer.chrome.com/docs/privacy-sandbox/attribution-reporting/understanding-noise/)

- `"aggregation_coordinator_origin"`
  - : The deployment option for the aggregation service.
- `"source_debug_key"` {{optional_inline}}
  - : A 64 bit unsigned integer representing the debugging key for the attribution source. This mirrors the value set in the associated {{httpheader("Attribution-Reporting-Register-Source")}} header's [`"debug_key"`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#debug_key) field. See [Debug reports](#debug_reports) for more information.
- `"trigger_debug_key"` {{optional_inline}}
  - : A 64 bit unsigned integer representing the debugging key for the attribution trigger. This mirrors the value set in the associated {{httpheader("Attribution-Reporting-Register-Trigger")}} header's `"debug_key"` field. See [Debug reports](#debug_reports) for more information.

## How noise is added to reports

> Noise plays a significant role in both event-level and summary reports as a method of protecting user privacy.

Reviewer: Hmmm - perhaps something more like this would be more accurate

> Noise is added to reports in order to obscure the trigger data in both event-level and summary reports, in order to protect user privacy.

IMO it would be worth explaining why this helps protect privacy too.

Reviewer: IN addition, the fixes further down are certainly more clear, but I'm still not certain at what point this noise becomes relevant.

I think I need a real example of trigger_data and trigger_data_matching, and how the noise is added, and how it then gets reported. Both here and in https://pr33172.content.dev.mdn.mozit.cloud/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Trigger#examples, and showing how the trigger and the source are match in https://pr33172.content.dev.mdn.mozit.cloud/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#examples

This `"trigger_data": [0, 1, 2, 3, 4]` does not make sense if you look at https://pr33172.content.dev.mdn.mozit.cloud/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Trigger#examples which has a big number in the trigger data, not an array like this

My "guess" is that somehow I am confusing how the browser matches the sources and events and how it reports them. So perhaps there is matching internally on trigger and source data, but this section is about what gets sent out in reports - so perhaps we perfectly match the source and trigger data, but we send out some reduced information about the trigger. 
Essentially I need to know about how the browser internally matches things and how that differs from the reports.

Does the question/confusion make sense?

Author: OK, I think your first sentence suggestion makes sense, so I've applied that change.

Author: I've read your other comment, and I'm not surprised you are confused. I don't understand it either. So what we need is:

1. A real example showing an `Attribution-Reporting-Register-Source` JSON containing `trigger_data` and `trigger_data_matching` fields
2. A corresponding `Attribution-Reporting-Register-Trigger` JSON containing `event_trigger_data`
3. An explanation of how noise is added in this situation.
4. An explanation of how these are matched with each other and how a report is sent out

@apasel422, can you provide anything like this? At the moment, Hamish and I are totally confused about how noise works.

Reviewer 2: Noise for event-level reports is not applied for trigger data, specifically, but rather for the entire output space of a source, which consists of the source's maximum number of attribubtions, the individual trigger data values permitted by the source's `trigger_data` field, and the source's report windows (how many windows there are and at what times they occur). This is done using [a randomized response algorithm](https://wicg.github.io/attribution-reporting-api/#obtaining-randomized-source-response). You can think of this as an algorithm like the following when a source is stored:

1. Generate the set of all possible sets of reports that can possibly originate from the source's configuration (including the set consisting of no reports).
2. Some percentage of the time (based on the cardinality of that set, the browser's implementation-specific privacy parameters, and the source's chosen `event_level_epsilon`), prevent the source from ever being attributed and instead pick a random member of that set to use as the source's reports.

Trigger-data matching works as follows:

In all cases, the source has an implicit or explicit set of allowed `trigger_data` values `allowedValues`. These are the only possible values that can appear in a report, regardless of the matching mode.

When a trigger with a particular `event_trigger_data.trigger_data` value `d` is attributed to that sources, the matching depends on the source's `trigger_data_matching`:

If it's `exact`, then `d` has to be a member of `allowedValues` in order for attribution to proceed. The trigger is rejected otherwise. In this case `d` will appear in the report as is.

If it's `modulus`, then `d` is taken modulus the size of `allowedValues`. In such a case, the value will always match, unless `allowedValues` was empty. This mode exists primarily for backwards compatibility with the API's behavior before `exact` was introduced, which was primarily for interoperability with the [Full Flexible Event-Level proposal](https://github.com/WICG/attribution-reporting-api/blob/main/flexible_event_config.md). When this mode is used, `d` itself isn't used, but rather `d % allowedValues.size`. This mode allows for some minor compression tricks by varying the `trigger_data` in the report based on modulus.

See https://github.com/WICG/attribution-reporting-api/issues/765 for background about the matching modes.

Unfortunately, this is quite complicated, so feel free to ask for further clarification.

Author: OK, I've read this several times and have updated the "How noise is added to reports" section of the "Generating reports" guide to suit. I think it makes a bit more sense now — I was under the impression that individual report data is modified in some way, but I think it is actually the case that some report outputs are modified based on the described algorithm, so the overall aggregate is noised, right?

@hamishwillee, please have a look and let me know if you think this makes any more sense. 

@apasel422 , can you give my new section a read through and let me know if it is technically correct?

Thanks folks.

Reviewer: @apasel422 This was really helpful. I missed it last time I reviewed.

> "Generate the set of all possible sets of reports that can possibly originate from the source's configuration (including the set consisting of no reports)."

So the main outstanding point I have now is down to what affects that set of generated reports. I think it is **just** the event numbers in the source right? 

This questions is captured in a little more detail here:
- https://github.com/mdn/content/pull/33172/files#r1613422329
- In https://github.com/mdn/content/pull/33172#discussion_r1615386037

## How noise is added to reports

> Noise is added to reports in order to obscure the output associated with a source, in order to protect user privacy. The exact source data cannot be identified and attributed back to individual users, but the overall patterns taken from the data will still provide the same meaning.

Reviewer: Love it. Suggestion because "in order" twice feels a _little_ unwieldy 
```suggestion
Noise is added to reports in order to obscure the output associated with a particular source, and thereby protect user privacy. The exact source data cannot be identified and attributed back to individual users, but the overall patterns taken from the data will still provide the same meaning.
```

Trigger data is limited quite strictly. By default, [navigation sources](/en-US/docs/Web/API/Attribution_Reporting_API/Registering_sources#navigation-based_attribution_sources) will be limited to only 3 bits of [`"trigger_data"`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#trigger_data) (the values 0 through 7), while [event sources](/en-US/docs/Web/API/Attribution_Reporting_API/Registering_sources#event-based_attribution_sources) will be limited to only 1 bit (the values 0 through 1). The `"trigger_data"` and [`"trigger_data_matching"`](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Source#trigger_data_matching) fields can be customized to vary the amount of noise applied. You could set a smaller range of trigger data to reduce the amount of noise applied, or a larger range to increase the noise:

Typical settings in the {{httpheader("Attribution-Reporting-Register-Source")}} header might look like so:

```json
{
  ...,
  // The following setting would reduce noise for a navigation
  // source (default 0–7), but increase noise for an event

>   // source (default 0–1)

Reviewer: This is unhelpful/gibberish. 
what do (default 0–7), mean?
Why do these settings reduce/increase noise?

Author: I don't know, and at this point, I just want to get this published so that someone can at least make use of some of it.

So I've just deleted the preceding paragraph that talks about this stuff and deleted these comments. As-is, the example is still fairly useful as an example of how a match can occur, but then the text explains that it might still not occur, due to the randomized response algorithm.

If lack of information on this becomes an issue later on, we can try to add something then.

Reviewer: @chrisdavidmills I do get this now and while deleting it is the right thing to do, this is actually what has clarified this all for me.

So what it is saying is that:

```
// default trigger_data for a navigation source is: [0, 1, 2, 3, 4, 5, 6, 7]
// default trigger_data for an event source is: [0, 1]
// The trigger_data we specify below of  [0, 1, 2, 3, 4] therefore has less noise that the default for a navigation source, and more noise than the default for an event source
```

The comment in the last line is a bit of a red herring. It implied to me that the trigger data is something you perhaps should play with to change the noise (I think you can, but that is not the point). It wasn't clear why the default values were as they are, and what the implications of this are.

The way I think it actually works is that the source defines what are possible valid data values in the registered trigger. It isn't stated, but I assume you can use any values you like and they don't have to be sequential

So consider we have two event values we want to actually trigger on: `[1,6]`. 
- We can set those values in trigger_data _if we want_ in the source
- We don't have to set those values for navigation triggers because they are in the default set
- if the source was an event source we would have to set the values (because 6 isn't in the default set)
- We have to set ONE of those values in the trigger_data in the trigger for matching

So what's all this about noise then? I THINK it is entirely down to the fact that the number of values you have in the trigger data changes the size of the potential report space. So if you just left the defaults you're saying that in some random number of cases above you might get a report for the event `4`, say, OR the event `6`, when the actual event was really `1`. The reporting server therefore can't know that any particular event is actually from the source it says it is, while still knowing that most data is broadly correct.

- I still don't understand how else reports might vary - i.e. is it just by event number or something else. This is the question we have here: https://github.com/mdn/content/pull/33172/files#r1613422329
- Is adding the noise something we WANT to do? I.e. should a developer be trying to increase the number of values in the source? 
- I don't know why navigation sources and event sources have different default ranges or what it implies for noise. My intuition is that this is important, but if you don't think it is ....

Reviewer 2: > I THINK it is entirely down to the fact that the number of values you have in the trigger data changes the size of the potential report space.

This is correct (as is the discussion preceding it in the same comment). The number of distinct values, rather than the values themselves, affects the size of the report space and thus noise.

> I still don't understand how else reports might vary

An individual event-level report is characterized by:

- Trigger data
- The time after source registration at which it is sent (determined by the source's event-level report windows and the time at which the trigger occurs)

A source, however, can produce multiple reports (determined by the source's `max_event_level_reports` field), so the output space for a *source* consists of some number of individual event-level reports (including 0 reports).

> Is adding the noise something we WANT to do? I.e. should a developer be trying to increase the number of values in the source?

It depends on the use case. If a developer needs to distinguish between `N` kinds of triggers, then it may make sense to have `N` trigger data values even if it increases noise (compared to `M < N`), because the true count of those kinds can be recovered using de-biasing.

> I don't know why navigation sources and event sources have different default ranges or what it implies for noise

The origin of this distinction is that event sources can be registered by a reporting origin without user interaction, whereas a navigation source can only be registered when a navigation occurs and there is transient user activation. As a result, event sources were restricted compared to navigation ones. However, at this point in the API's evolution, trigger data is completely customizable for navigation sources and event sources (up to 32 distinct values), and noise is characterized by the size of the output space in addition to the source's `event_level_epsilon`, so the original distinction doesn't matter much anymore (beyond backwards compatibility), since the noising computations work the same for both kinds of sources.

Reviewer: @chrisdavidmills What I get from this that I am _fairly_ sure of is:

- Source trigger data has up to 32 event values.
  - These map to actual events and random values that you can use to add noise
  - The different default values for the navigation/event sources don't matter any more.
  - Not clear if they have to be sequential and 0 indexed or whatever, but it probably doesn't matter.
- Event trigger data value must be a value in source trigger data for attribution.
- The size of the possible report space is affected by amount of source trigger data and also the reporting windows.
  - It sounds like it might also be also _capped_ by `max_event_level_reports`, though that isn't certain. 
  - My guess is that you probably have "other reasons" for setting your max number of reports and your reporting windows, so if you want to mess with the noise, changing the number of source trigger event values is the easiest way that has no other side-effects.

> It depends on the use case. If a developer needs to distinguish between N kinds of triggers, then it may make sense to have N trigger data values even if it increases noise (compared to M < N), because the approximate count of those kinds can be recovered using de-biasing.

I don't understand this (at all - especially the bit on de-biassing). If you need to distinguish between N trigger events then how can you do that without having distinct events - if you don't have N events then you won't have N possible attributions in the report so there is no way to distinguish.

Or to put it another way, I would have though you'd always need at least as many values in the array as you have events you want to track. The question is more "when would I want more"?
It certainly doesn't answer the question "do we want to add noise, and when", but it strongly implies that you do sometimes want to play with the noise.

Anyway, I think it is worth noting that there are 32 values. It is worth note the factors that affect the report space, and that trigger data is the safest/lowest impact way ... assuming I am right.

@apasel422 Thanks for your answer. I think I just don't know enough to be sure of the noise implications. Perhaps @chrisdavidmills will get more from this.

  "trigger_data": [0, 1, 2, 3, 4],

>   // Could also be "modulus"

Reviewer: We need to forward link or otherwise explain that most of the time they have to specify exact.

Author: This comment is now deleted

> }

Reviewer: So this doesn't match what I see inhttps://pr33172.content.dev.mdn.mozit.cloud/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Trigger#examples

Author: I think this is trigger data specified on the source side, not the trigger side.
